sycl/include/CL/sycl/interop_handle.hpp

```diff
@@ -0,0 +1,117 @@
+//==------------ interop_handle.hpp --- SYCL interop handle ----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/access/access.hpp>
+#include <CL/sycl/detail/accessor_impl.hpp>
+#include <CL/sycl/detail/common.hpp>
+#include <CL/sycl/detail/defines.hpp>
+#include <CL/sycl/detail/pi.hpp>
+
+#include <memory>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+
+namespace detail {
+class AccessorBaseHost;
+class ExecCGCommand;
+class DispatchHostTask;
+} // namespace detail
+
+template <typename DataT, int Dims, access::mode AccMode,
+          access::target AccTarget, access::placeholder isPlaceholder>
+class accessor;
+
+class interop_handle {
+public:
+  /// Receives a SYCL accessor that has been defined is a requirement for the
+  /// command group, and returns the underlying OpenCL memory object that is
+  /// used by the SYCL runtime. If the accessor passed as parameter is not part
+  /// of the command group requirements (e.g. it is an unregistered placeholder
+  /// accessor), the exception `cl::sycl::invalid_object` is thrown
+  /// asynchronously.
+  template <typename dataT, int dimensions, access::mode accessmode,
+            access::target accessTarget, access::placeholder isPlaceholder>
+  typename std::enable_if<accessTarget != access::target::host_buffer,
+                          cl_mem>::type
+  get_native_mem(const accessor<dataT, dimensions, accessmode, accessTarget,
+                                isPlaceholder> &Acc) const {
+#ifndef __SYCL_DEVICE_ONLY__
+    // employ reinterpret_cast instead of static_cast due to cycle in includes
+    // involving CL/sycl/accessor.hpp
+    auto *AccBase = const_cast<detail::AccessorBaseHost *>(
```

Comment:
I'd like to avoid all this if we can. Could you please explain why it is necessary? I don't understand the comment above it.

Reply:
> I'd like to avoid all this if we can

What exactly would you like to avoid?

Please, refer to [inheritance diagram of accessor on host side](https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/accessor.hpp#L61).
As you see `accessor` type is derivative of `AccessorBaseHost` which is also named as  `Requirement`. In implementation we will need to include `interop_handle.hpp` in `CL/sycl/detail/cg.hpp` which will produce a cycle if we include `accessor.hpp` here.

Another approach is to move `interop_handle` class declaration to `CL/sycl/detail.cg.hpp` file. Though, somehow I can't see a viable reason for this decision to be correct.

Reply:
I'm not a fan of using `reinterpret_cast` if we can avoid it either. I don't quite get the cycle we're going to get if we include `accessor.hpp` here (and indirectly in `cg.hpp`). Could you elaborate a bit on that?

Reply:
> What exactly would you like to avoid?

I am concerned with the use of `reinterpret_cast` and `const_cast` in this code. With `const_cast`, the issues are simple. `const` is a lovely feature that helps us write correct code, and [casting it away is almost always indicative of a code design problem](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es50-dont-cast-away-const). I think we can easily change interfaces or introduce `const` in a few places to avoid its use. However, the `reinterpret_cast` use is more concerning.

This code is reinterpreting a pointer to a forward declaration of a child class to a pointer to its parent class. The only reason it works is because `accessor` inherits `AccessorBaseHost` first, and they are both standard layout. If anyone changes the layout of `accessor` in the future, [then the code could easily break](https://godbolt.org/z/j3Hxw3), which would require substantial debugging to diagnose. A `static_cast` will ensure that the pointer is offset correctly when upcasting. With `reinterpret_cast`, the same address will be used and the compiler will simply treat the memory there as being of a different type. Clang should warn about this, but perhaps it cannot because the code is reintepreting a pointer to a foward declaration.

I am happy to work with you to come up with an alternative solution to this issue.

> which will produce a cycle if we include accessor.hpp here.

I'm afraid I don't see this circular include you're referring to. Can you please elaborate?

Reply:
> which will produce a cycle if we include accessor.hpp here

Suppose we add `#include <CL/sycl/accessor.hpp> in `interop_handle.hpp`. Furthermore we will add `#include <CL/sycl/interop_handle.hpp>` to `include/CL/sycl/handler.hpp`. Then, we will have the following cycle (here `->` says `which includes`):

```
handler.hpp -> interop_handle.hpp -> accessor.hpp -> handler.hpp (cycle here)
```

This cycle will produce compile-time errors like `invalid use of incomplete type ‘class cl::sycl::handler’`.

I'll try to eliminate this cycle and will put it into distinct patch.

Reply:
See https://github.com/intel/llvm/pull/1770

Reply:
I don't think that PR is necessary. The cycle you've identified should be easily broken by include guards that prevent the same header being included more than once. Did you try and compile it? You should be able to include `accessor.hpp` in `interop_handle.hpp` to no ill effects, enabling the use of `static_cast` instead of `reinterpret_cast`.

Having looked at the code more closely, I am now confused as to why a `const_cast` was necessary at all. The `detail::getSyclObjImpl` function takes the accessor base object a reference to `const AccessorBaseHost`, meaning you are casting away the const only for it to be re-applied when it is passed as an argument. 

I think this code could be easily simplified by including `accessor.hpp` and replacing the `reinterpret_cast` and `const_cast` with a single `static_cast`.

```cpp
  template <backend BackendName = backend::opencl, typename DataT, int Dims,
            access::mode Mode, access::target Target, access::placeholder IsPlh>
  typename std::enable_if<
      Target != access::target::host_buffer,
      typename interop<BackendName,
                       accessor<DataT, Dims, Mode, Target, IsPlh>>::type>::type
  get_native_mem(const accessor<DataT, Dims, Mode, Target, IsPlh> &Acc) const {
#ifndef __SYCL_DEVICE_ONLY__
    const auto *AccBase = static_cast<const detail::AccessorBaseHost *>(&Acc);
    return getMemImpl<BackendName, DataT, Dims, Mode, Target, IsPlh>(
        detail::getSyclObjImpl(*AccBase).get());
#else
    (void)Acc;
    return nullptr;
#endif
  }
```

Reply:
> The cycle you've identified should be broken by include guards that prevent the same header being included more than once. Did it fail to compile for you? This code should be able to include accessor.hpp in interop_handle.hpp to no ill effects, enabling the use of static_cast instead of reinterpret_cast.

The cycle I've identified results in compile errors saying `invalid use of incomplete type ‘class cl::sycl::handler’` in `accessor.hpp` due to the following include chain:

```
handler.hpp (pragma once, do the include, though no handler definition seen yet) -> interop_handle.hpp -> accessor.hpp -> handler.hpp (pragma once, can't really include)
```

> The const_cast in this code is unnecessary.

Sure thing, it's redundant

Reply:
That is curious to me, could you please share your compiler output?

Reply:
I bet you're able to see this failure now in buildbot.

Reply:
I believe the issue here is resolved now.

Reply:
Curious, I wasn't able to reproduce this locally.